Source/Core/VideoBackends/OGL/ProgramShaderCache.cpp

```diff
@@ -502,88 +605,253 @@ void ProgramShaderCache::Init()
+    if (!s_async_compiler->HasWorkerThreads())
     {
-      ERROR_LOG(VIDEO, "GL_ARB_get_program_binary is supported, but no binary format is known. So "
-                       "disable shader cache.");
-      g_ogl_config.bSupportsGLSLCache = false;
+      // No point using the async compiler without workers.
+      s_async_compiler.reset();
     }
-    else
-    {
-      if (!File::Exists(File::GetUserPath(D_SHADERCACHE_IDX)))
-        File::CreateDir(File::GetUserPath(D_SHADERCACHE_IDX));
+  }
+
+  // Read our shader cache, only if supported and enabled
+  if (g_ogl_config.bSupportsGLSLCache && g_ActiveConfig.bShaderCache)
+    LoadProgramBinaries();
 
-      std::string cache_filename =
-          StringFromFormat("%sogl-%s-shaders.cache", File::GetUserPath(D_SHADERCACHE_IDX).c_str(),
-                           SConfig::GetInstance().GetGameID().c_str());
+  CreateHeader();
 
-      ProgramShaderCacheInserter inserter;
-      g_program_disk_cache.OpenAndRead(cache_filename, inserter);
-    }
-    SETSTAT(stats.numPixelShadersAlive, pshaders.size());
+  CurrentProgram = 0;
+  last_entry = nullptr;
+  last_uber_entry = nullptr;
+
+  if (g_ActiveConfig.CanPrecompileUberShaders())
+    PrecompileUberShaders();
+}
+
+void ProgramShaderCache::RetrieveAsyncShaders()
+{
+  if (s_async_compiler)
+    s_async_compiler->RetrieveWorkItems();
+}
+
+void ProgramShaderCache::Reload()
+{
+  if (s_async_compiler)
+  {
+    s_async_compiler->WaitUntilCompletion();
+    s_async_compiler->RetrieveWorkItems();
   }
 
-  CreateHeader();
+  const bool use_cache = g_ogl_config.bSupportsGLSLCache && g_ActiveConfig.bShaderCache;
+  if (use_cache)
+    SaveProgramBinaries();
+
+  s_program_disk_cache.Close();
+  s_uber_program_disk_cache.Close();
+  DestroyShaders();
+
+  if (use_cache)
+    LoadProgramBinaries();
+
+  if (g_ActiveConfig.CanPrecompileUberShaders())
+    PrecompileUberShaders();
 
+  InvalidateVertexFormat();
   CurrentProgram = 0;
   last_entry = nullptr;
+  last_uber_entry = nullptr;
+  last_uid = {};
+  last_uber_uid = {};
 }
 
 void ProgramShaderCache::Shutdown()
 {
+  if (s_async_compiler)
+  {
+    s_async_compiler->WaitUntilCompletion();
+    s_async_compiler->StopWorkerThreads();
+    s_async_compiler->RetrieveWorkItems();
+    s_async_compiler.reset();
+  }
+
   // store all shaders in cache on disk
-  if (g_ogl_config.bSupportsGLSLCache)
+  if (g_ogl_config.bSupportsGLSLCache && g_ActiveConfig.bShaderCache)
+    SaveProgramBinaries();
+  s_program_disk_cache.Close();
+  s_uber_program_disk_cache.Close();
+
+  InvalidateVertexFormat();
+  DestroyShaders();
+  s_buffer.reset();
+}
+
+void ProgramShaderCache::BindVertexFormat(const GLVertexFormat* vertex_format)
+{
+  u32 new_VAO = vertex_format ? vertex_format->VAO : 0;
+  if (s_last_VAO == new_VAO)
+    return;
+
+  glBindVertexArray(new_VAO);
+  s_last_VAO = new_VAO;
+}
+
+void ProgramShaderCache::InvalidateVertexFormat()
+{
+  s_last_VAO = static_cast<u32>(-1);
```

Comment:
Never seen this approach before, is this to overflow to the max value?  I feel like it'd be nicer to use something more explicit.